### ISSUE TITLE
Unclear corner cases about newer versions

### ISSUE THREAD

**Issue by krOoze** [KhronosGroup/Vulkan-Docs#1097]
So, there are three sources of Vulkan version: `vkEnumerateInstanceVersion::pApiVersion`, `VkApplicationInfo::apiVersion`, and `VkPhysicalDeviceProperties::apiVersion`.

1) `VkApplicationInfo::apiVersion` gives bit of a weird message here:

> Only the major and minor versions of the instance **must** match those requested in `apiVersion`.

It feels bit unclear who is addressed here and what is it saying. I read it 10x and I think it is just weird wording. Probably should be:

> The implementation will create an instance matching the major and minor version requested in `apiVersion`, or if that is not supported then the highest version the instance does support.

2) There is bit of a problem with layers:

> Implicit layers **must** be disabled if they do not support a version at least as high as `apiVersion`.

That is interesting. It feels something like that should apply to explicit layers too. Or is it like at users own risk? Then at least Note should be there.

Also it is unclear if patch version number is ignored here too, or not.

3) It is ambiguous how `vkGetInstanceProcAddr` works.

Will it return `NULL` for 1.1 commands if `VkApplicationInfo::apiVersion` is 1.0?

If the instance is 1.1 but physical devices are 1.0, what will it return for 1.1 device commands?

4) It is unclear how extension core version requirement works

E.g. `VK_KHR_spirv_1_4` has "Requires Vulkan 1.1". It is a device extension. It is unclear which of the three Vulkan versions it references (or if all of them must be at least 1.1).

Furthemore it is not clear how it is checked. `vkCreateInstance` VU only covers extension-extension dependencies. Does version dependency cause `VK_ERROR_EXTENSION_NOT_PRESENT` or is it an undefined behavior?

5) There is bit of a catch-22 with `vkGetPhysicalDeviceProperties2`:

> Physical-device-level functionality or behavior added by a new core version of the API **must** not be used unless it is supported by the physical device as determined by `VkPhysicalDeviceProperties::apiVersion` and the specified version of `VkApplicationInfo::apiVersion`.

It feels 1.1 instance should emulate at least this command.

**Comment by critsec** [comment]
1. I agree this is weird, and probably left over from before 1.1. It should probably be moved into the next sentence about 1.0 implementations, something like "Vulkan 1.0 implementations were required to return `VK_ERROR_INCOMPATIBLE_DRIVER` if the major and minor portion of `apiVersion` was larger than 1.0.".

Note that `VkApplicationInfo::apiVersion` is *never* required to change the behavior of the loader or driver, except for the sentence I just mentioned on 1.0 implementations. So saying that the created instance is somehow different based on this value wouldn't be right. Starting in 1.1, this version is meant for use by validation layers and similar.

---

2. The application has no control over implicit layers, so this is saying that the loader can't allow an inconsistent/incompatible configuration. The application has control over explicit layers, and presumably knows (and the developer has tested) what it's enabling. I do think there's a mistake here in that it should only disable layers that don't support `min(VkApplicationInfo::apiVersion, instance version)`.

---

3. The loader and driver can ignore `VkApplicationInfo::apiVersion`, so it doesn't affect what `vkGetInstanceProcAddr` returns; that should only be based on what the loader/drivers support. So for your specific scenarios:

* If `vkEnumerateInstanceVersion` is 1.1, then `vkGetInstanceProcAddr` should return a valid function pointer for any 1.1 core API command, regardless of `VkApplicationInfo::apiVersion`.

* If the application calls through a function pointer for a 1.1 device command with a 1.0 dispatchable object, behavior is undefined. At the instance level, since you can have a mix of 1.0 and 1.1 physical devices, the loader has to support returning a function pointer that will work when called with a 1.1 device but will behave in an undefined way when called with a 1.0 device. The case where you only have 1.0 physical devices isn't really different from the loader's point of view.

---

4. Yeah, this isn't clear in the spec, we should clarify that. I think the only thing that makes sense is for it to refer to physical device version. The loader and driver aren't supposed to be checking that dependencies are satisfied, but in your example a driver (physical device) that doesn't support Vulkan 1.1 really shouldn't be advertising extensions that require Vulkan 1.1: that would be a driver bug.

---

5. Yeah, this one is a pet peeve of mine :). The only valid way to do it is to first call `vkGetPhysicalDeviceProperties` on a physical device to get its supported version, and then if it is >= 1.1 call `vkGetPhysicalDeviceProperties2`. IMO the best way to fix that would be to add a new `vkEnumeratePhysicalDevices2` that returns a physical device version with each physical device handle. That would be a new instance-level core API call and could be implemented entirely in the loader (internally calling `vkGetPhysicalDeviceProperties` for physical devices that don't support the new instance version). This hasn't been a priority for the WG to fix though: it's a mild annoyance, but not a significant functionality or performance limitation.

**Comment by krOoze** [comment]
> Note that `VkApplicationInfo::apiVersion` is never required to change the behavior of the loader or driver, except for the sentence I just mentioned on 1.0 implementations.

Well, there are some things in the spec that could be interpreted that way. E.g. the `vkGetInstanceProcAddr` (point nr. 3).

> I do think there's a mistake here in that it should only disable layers that don't support min(VkApplicationInfo::apiVersion, instance version).

The spec explicitly points out the physical device version could be higher than instance version. So that could break for `VkApplicationInfo::apiVersion=1.1`, `vkEnumerateInstanceVersion=1.0`, and `VkPhysicalDeviceProperties::apiVersion=1.1`.

So actually `min(app version, max(instance version, device version))`?

> If `vkEnumerateInstanceVersion` is 1.1, then `vkGetInstanceProcAddr` should return a valid function pointer for any 1.1 core API command, regardless of `VkApplicationInfo::apiVersion`.

Hm, could use a clarifying tweak in the table. Because disabled extensions **do** return `NULL`. It is inconsistent logic that "disabled" versions would return fptrs.

> but in your example a driver (physical device) that doesn't support Vulkan 1.1 really shouldn't be advertising extensions that require Vulkan 1.1: that would be a driver bug.

Well, it is not exactly clear either if the Requirement also addresses Implementers. Maybe it does; IIRC the Device Extensions are indeed filtered out if the Instance Extension dependency is not supported by the Instance\Loader.

Though e.g. the `VK_KHR_spirv_1_4` is published even in the Vulkan 1.0 spec, despite the 1.1 requirement.

>  IMO the best way to fix that would be to add a new vkEnumeratePhysicalDevices2 that returns a physical device version with each physical device handle. That would be a new instance-level core API call and could be implemented entirely in the loader

Well, that would be somewhat consistent with `vkEnumerateInstanceVersion`.

**Comment by krOoze** [comment]
added nr. 6

**Comment by Rua** [comment]
> > If `vkEnumerateInstanceVersion` is 1.1, then `vkGetInstanceProcAddr` should return a valid function pointer for any 1.1 core API command, regardless of `VkApplicationInfo::apiVersion`.
> 
> Hm, could use a clarifying tweak in the table. Because disabled extensions **do** return `NULL`. It is inconsistent logic that "disabled" versions would return fptrs.

I agree, I didn't even realise that Vulkan worked this way. Maybe it could be clarified in the documentation of `VkApplicationInfo::apiVersion` that it has no effect on the behaviour of any Vulkan functions, whether from instance or driver, and affects only what usage is valid and what isn't. Also, I believe there's no reason not to set this to a really high value (say, Vulkan 99999.0), and have automatic support for all current and future Vulkan versions? What is even the point of this value then?
